#!/bin/bash

#SBATCH --time=24:00:00
#SBATCH --account=def-coling
#SBATCH --mail-type=ALL
#SBATCH --mail-user=edegreef@ucdavis.edu
#SBATCH --nodes=1
#SBATCH --ntasks=1
#SBATCH --mem=0
#SBATCH --cpus-per-task=24
#SBATCH --kill-on-invalid=yes
#SBATCH --job-name=platypus
#SBATCH --output=%x-%j.out

module load nixpkgs/16.09  gcc/7.3.0 platypus/0.8.1

# Run platypus to call variants from bams & reference genome 
# Make sure all bam files are indexed beforehand. Also couldn't add Platypus.py to path for some reason so I inputed the path here in script
python /cvmfs/soft.computecanada.ca/easybuild/software/2017/avx2/Compiler/gcc7.3/platypus/0.8.1/bin/Platypus.py callVariants \
--bamFiles=/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/HamLB03-1.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/HamLB03-2.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/HamLB03-3.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/Hyam-2016-06-pooled.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-12-710-509.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-13-702-507.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-15-708-507.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-16-705-509.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2016-02-706-509.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2016-05-715-506.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2016-06.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2016-09.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2016-10-706-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2016-11.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2016-12.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2016-13-707-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2016-14.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2016-17.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2017-01.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/downsampled_bams/NBW-2017-03.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2017-06.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2017-2-705-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2017-4-704-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2019-01-709-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/NBW_Dec2020/NBW-2019-02-714-509.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW19-03-712-504.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW1904-718-503.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW1906-710-504.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW1908-711-504.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW-2019-B3-716-504.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW-2019-B10-718-504.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW-2019-B11-718-505.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW-2019-B12-710-505.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW-2019-B13-711-505.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW-2019-B14-712-505.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW-2019-B15-712-503.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/NBW-2019-B16-717-503.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/NBW-2019-B17-716-503.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-01-711-503.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-03-717-504.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-04-712-502.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-05-717-502.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/merged_bams/Ha14-06-716-502-2.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/Ha14-07-718-502.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-08-710-503.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/merged_bams/Ha14-10-711-502-2.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/Ha14-13-716-505.deDupRG.downsampled.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/Ha14-16-710-502.deDupRG.bam,/scratch/edegreef/NBW/aligned_bams/dedupRG/downsampled_bams/Ha14-18-717-505.deDupRG.downsampled.bam \
--refFile=/scratch/edegreef/ref_genome/Northern_bottlenose_whale_051018_shortLabel.fasta \
--output=/scratch/edegreef/NBW/NBW2_allvariantcalls_2mergedbam.vcf \
--nCPU=24
